### ISSUE TITLE
`many/p` does not backtrack on trailing separator

### ISSUE THREAD

**Issue by lihebi** [lexi-lambda/megaparsack#27]
I'm trying to match some repeated `a` separated by `.` and with a final `.`, for example `a.a.a.`.

The following code:

```racket
  (parse-result! (parse-string (do (many/p (char/p #\a)
                                           #:sep (char/p #\.))
                                   (char/p #\.))
                               "a.a.a."))
```

Does not match, with error:

```
; parse error
;   unexpected: end of input
;   expected: 'a'
```

I think the problem is that, when it sees the final `.`, it tries to interpret it as a separator and tries to match `a`. Upon failure, should it backtrack and match the final `.`?

Adding `try/p` around the patterns does not solve the issue.

**Comment by lexi-lambda** [comment]
I don’t think making `many/p` implicitly add `try/p` around the entire optional part of the parser is the right solution here, since that’s a very heavy hammer, and it might destroy parse errors. I suspect you probably don’t want to use `#:sep` here at all. Try this, instead:

```racket
(many/p (do (char/p #\a) (char/p #\.)))
```

Now there’s no ambiguity.

**Comment by lihebi** [comment]
> ```racket
> (many/p (do (char/p #\a) (char/p #\.)))
> ```

Thanks. This works for this example. But my use case back then was more complicated than this example, and I remember I really wanted a `#:sep`. Let me try to retrieve my memory ..

**Comment by sorawee** [comment]
This issue was brought up on Discord. The workaround here doesn't work for the use case that we have. In particular, we want to parse a list of integers separated by exactly one space, followed by possible trailing spaces, then followed by `x`. The naive attempt is:

```
(define parse-all/p
  (do (many/p integer/p #:sep space/p)
    (many/p space/p)
    (char/p #\x)
    (pure #t)))

;; should parse
(parse-string parse-all/p "41 48     x") ; (A)
(parse-string parse-all/p "41 48x") ; (B)

;; should not parse
(parse-string parse-all/p "41     48x") ; (C)
```

But this fails exactly as this issue reported. 

We might want to switch `(many/p integer/p #:sep space/p)` to `(many/p (do integer/p space/p))`, but this would violate (B). 

We also might want to switch `(many/p integer/p #:sep space/p)` to `(many/p (do integer/p (many/p space/p)))`, but this would violate (C).

**Comment by AlexKnauth** [comment]
Letting the C string parse successfully (edit: by switching to `(many/p (do integer/p (many/p space/p)))`) seems fine to me.

But, if that's truly what you want, then a workaround like this might help:
```racket
(define (workaround-many/p parser #:sep sep)
  (many/p parser #:sep (try/p (do sep (lookahead/p parser)))))
```
Edit: see also comment below with a better way to do this

On the original issue:
```racket
(parse-result! (parse-string (do (workaround-many/p (char/p #\a)
                                                    #:sep (char/p #\.))
                                 (char/p #\.))
                             "a.a.a."))
```

And on your example:
```racket
(define parse-all/p
  (do (workaround-many/p integer/p #:sep space/p)
    (many/p space/p)
    (char/p #\x)
    (pure #t)))
```

**Comment by lexi-lambda** [comment]
@sorawee The issue with your “naive attempt” is that the parsed passed as an argument to `#:sep` successfully eats a space, and successfully making progress is precisely what normally causes a parser to commit. Fortunately, the solution is quite simple, which is to add a use of [`noncommittal/p`](https://docs.racket-lang.org/megaparsack/reference.html#%28def._%28%28lib._megaparsack%2Fmain..rkt%29._noncommittal%2Fp%29%29):

```racket
(define parse-all/p
  (do (many/p integer/p #:sep (noncommittal/p space/p))
      (many/p space/p)
      (char/p #\x)
      (pure #t)))
```

Now your examples all parse as desired:

```
> (parse-string parse-all/p "41 48     x")
(success #t)
> (parse-string parse-all/p "41 48x")
(success #t)
> (parse-result! (parse-string parse-all/p "41     48x"))
string:1:7: parse error
  unexpected: 4
  expected: 'x' or whitespace
```

**Comment by lexi-lambda** [comment]
…in fact, I’m now realizing that this issue is essentially the same exact thing that prompted me to add `noncommittal/p` last year in the first place. So I think it can now be safely closed as resolved by 0ccdee4270da0337700ac62aa106735d0d879695.
